### PR TITLE
Get parameters information of operator backed services from Kubernetes OpenAPI spec

### PR DESCRIPTION
**Pull request by feloy**
**What type of PR is this?**

/kind feature

**What does this PR do / why we need it**:

`odo catalog describe service <service/crd>`:
- downloads the OpenAPI v2 (swagger) definition from the Kubernetes API
- gets the Spec of the CRD as an OpenAPI v2 schema
- displays the OpenAPI v2 schema as is for the JSON output
- displays theOpenAPI v2 schema in a tabular form for the text output
- displays if required and the type of each parameter on text output
- If the OpenAPI Spec is empty for the CRD, the information from the CSV is transformed in OpenAPI v2 Spec format

Not covered by this PR:
- `odo service create` only checks parameters from CSV, not OpenAPI v2 spec
- Downloaded OpenAPI v2 spec is not cached

**Which issue(s) this PR fixes**:

Fixes part of #4850

**PR acceptance criteria**:

- [x] Unit test 

- [x] Integration test 

- [ ] Documentation 

- [x] I have read the [test guidelines](https://github.com/openshift/odo/blob/master/docs/dev/test-architecture.adoc)

**How to test changes / Special notes to the reviewer**:

### CODE REVIEW

**Comment by dharmit** [comment]
>  * If the OpenAPI Spec is empty for the CRD, the information from the CSV is transformed in OpenAPI v2 Spec format

I installed an Operator that had neither information available:
```sh
# https://operatorhub.io/operator/postgres-operator
$ kubectl create -f https://operatorhub.io/install/postgres-operator.yaml
```
and `odo catalog service describe` output looked like:
```sh
$ odo catalog describe service postgres-operator.v1.4.0/postgresql        
Kind: postgresql
Version: v1
Description: Creates a PostgreSQL cluster managed by Postgres-Operator.
Parameters:

 REQUIRED  PATH  TYPE  DISPLAYNAME  DESCRIPTION
```
This is also how odo built from `main` branch behaves when it can't find anything in `x-descriptors` field of the CSV. For example PTAL at https://github.com/openshift/odo/issues/4850#issuecomment-900881020.

I think odo should print something useful in the cases where an Operator doesn't provide any info at either places instead of showing an empty table like it does right now. The reason I think that way is because if we continue with empty table output, it'll make the user think that odo couldn't provide them with the info. odo could clearly say that the Operator in question doesn't provide any usable info.

WDYT?

**Comment by feloy** [comment]
> > * If the OpenAPI Spec is empty for the CRD, the information from the CSV is transformed in OpenAPI v2 Spec format
> 
> I installed an Operator that had neither information available:
> 
> [...]
> This is also how odo built from `main` branch behaves when it can't find anything in `x-descriptors` field of the CSV. For example PTAL at [#4850 (comment)](https://github.com/openshift/odo/issues/4850#issuecomment-900881020).
> 
> I think odo should print something useful in the cases where an Operator doesn't provide any info at either places instead of showing an empty table like it does right now. The reason I think that way is because if we continue with empty table output, it'll make the user think that odo couldn't provide them with the info. odo could clearly say that the Operator in question doesn't provide any usable info.
> 
> WDYT?

You are right, it is interesting to know the data is not available. I added a message.

**Comment by dharmit** [comment]
/approve

Functionality looks good. Got few questions; will lgtm after they're resolved.

File: pkg/kclient/operators.go [hunk 1]
```diff
@@ -146,3 +148,59 @@ func (c *Client) GetCSVWithCR(name string) (*olm.ClusterServiceVersion, error) {
...
+				crd = definition
+				found = true
+				break loopDefinitions
+			}
+		}
+
+	}
+	if !found {
+		return nil, errors.New("no definition found")
+	}
```
[the remark this answers is not in the document]
Comment (feloy) [reply pkg/kclient/operators.go:198]: We are looking for the <cr> definition here

File: pkg/kclient/operators.go [hunk 2]
```diff
@@ -146,3 +148,59 @@ func (c *Client) GetCSVWithCR(name string) (*olm.ClusterServiceVersion, error) {
...
+
+	spec, ok := crd.Properties["spec"]
+	if ok {
+		return &spec, nil
+	}
```
[the remark this answers is not in the document]
Comment (feloy) [reply pkg/kclient/operators.go:203]: More exactly, they could be no `spec` field. In this case, there is no useful information to retrieve

File: pkg/odo/cli/catalog/describe/operator_backend.go [hunk 3]
```diff
@@ -77,6 +83,17 @@ func (ohb *operatorBackend) ValidateDescribeService(dso *DescribeServiceOptions)
...
+	crd, err := dso.KClient.GetResourceSpecDefinition(ohb.CR.Name, ohb.CR.Version, ohb.CustomResource)
+	if err != nil {
+		log.Warning("Unable to get CRD specifications:", err)
+		return nil
+	}
+	ohb.CRDSpec = crd
+
+	if crd == nil {
+		ohb.CRDSpec = toOpenAPISpec(cr)
+	}
```
[the remark this answers is not in the document]
Comment (feloy) [reply pkg/odo/cli/catalog/describe/operator_backend.go:96]: I find more readable to first work by block. The first block is about CRD. If the result is empty, we go the the second block, about CSV. Your solution mixes the 2 blocks (we finally affect the CRD to CRDSpec after we work with CSV)

File: pkg/kclient/operators.go [hunk 4]
```diff
@@ -146,3 +148,59 @@ func (c *Client) GetCSVWithCR(name string) (*olm.ClusterServiceVersion, error) {
 	}
 	return &olm.ClusterServiceVersion{}, fmt.Errorf("could not find any Operator containing requested CR: %s", name)
 }
+
+// GetResourceSpecDefinition returns the OpenAPI v2 definition of the Kubernetes resource of a given group/version/kind
+func (c *Client) GetResourceSpecDefinition(group, version, kind string) (*spec.Schema, error) {
```
[the remark this answers is not in the document]
Comment (feloy) [reply pkg/kclient/operators.go:153]: Yes, I'll write some unit tests for this function

File: pkg/odo/cli/catalog/describe/operator_backend.go [hunk 5]
```diff
@@ -108,30 +125,108 @@ func (ohb *operatorBackend) RunDescribeService(dso *DescribeServiceOptions) erro
...
-	fmt.Printf("Version: %s\n", repr.Version)
-	fmt.Printf("Description: %s\n", repr.Description)
-	fmt.Println("Parameters:")
+func HumanReadableOutput(w io.Writer, service service.OperatorBackedService) {
+	fmt.Fprintf(w, "Kind: %s\n", service.Spec.Kind)
+	fmt.Fprintf(w, "Version: %s\n", service.Spec.Version)
+	fmt.Fprintf(w, "Description: %s\n", service.Spec.Description)
 
+	if service.Spec.Schema == nil {
+		log.Warning("Unable to get parameters from CRD or CSV")
```
Comment (dharmit) [inline pkg/odo/cli/catalog/describe/operator_backend.go:143]:
```suggestion
		log.Warningf("Unable to get parameters from CRD or CSV; Operator %q doesn't have the required information", service.Name)
```

Comment (dharmit) [reply pkg/odo/cli/catalog/describe/operator_backend.go:143]: @feloy it's probably just me, but I feel like being explicit about pointing why odo couldn't get the parameters for the user.

File: pkg/odo/cli/catalog/describe/operator_backend.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (dharmit) [reply pkg/odo/cli/catalog/describe/operator_backend.go:96]: Sure, makes sense. Not "questioning" your preference here, but is this something you came across while reading "Learning Go" or similar book or is it just a personal preference?

[the remark this answers is not in the document]
Comment (feloy) [reply pkg/odo/cli/catalog/describe/operator_backend.go:96]: I think it is specific to and idiomatic in Go, similarly to how are handled errors, by first handling the errors and at the end the happy path. 
Also, Go tries to keep the contexts as small as possible. 

As an aside, I think I've read in "Learning Go" that the lengths of variable names are proportional to the size of the contexts they are in: Aan`i` variable is good in a small context, and `aVeryVerboseName` is needed in a long function. But long functions/contexts are discouraged in Go, so are preferred short variable names :)

**Comment by dharmit** [comment]
Doesn't seem like a failure related to the PR

/test v4.8-integration-e2e

```
 • Failure [48.517 seconds]
odo devfile push command tests
/go/src/github.com/openshift/odo/tests/integration/devfile/cmd_devfile_push_test.go:18
  when Create and push java-springboot component
  /go/src/github.com/openshift/odo/tests/integration/devfile/cmd_devfile_push_test.go:779
    should execute default build and run commands correctly [It]
    /go/src/github.com/openshift/odo/tests/integration/devfile/cmd_devfile_push_test.go:789
    Unexpected error:
      
        <*errors.errorString | 0xc00047c600>: {
            s: "cmd [ps -ef] failed with error Signal 23 (URG) caught by ps (3.3.15).\nps:ps/display.c:66: please report this bug\ncommand terminated with exit code 1\n on pod fkthlt-app-7497874494-7rjtt",
        }
        cmd [ps -ef] failed with error Signal 23 (URG) caught by ps (3.3.15).
        ps:ps/display.c:66: please report this bug
        command terminated with exit code 1
         on pod fkthlt-app-7497874494-7rjtt
    occurred
    /go/src/github.com/openshift/odo/tests/integration/devfile/cmd_devfile_push_test.go:813 
```

**Pull request merged**